pkg/kyverno/common/fetch.go

```diff
@@ -19,49 +21,86 @@ import (
 // the resources are fetched from
 // - local paths to resources, if given
 // - the k8s cluster, if given
-func GetResources(policies []*v1.ClusterPolicy, resourcePaths []string, dClient *client.Client) ([]*unstructured.Unstructured, error) {
-	var resources []*unstructured.Unstructured
+func GetResources(policies []*v1.ClusterPolicy, resourcePaths []string, dClient *client.Client, cluster bool, namespace string) ([]*unstructured.Unstructured, error) {
+	resources := make([]*unstructured.Unstructured, 0)
 	var err error
+	var resourceTypesMap = make(map[string]bool)
+	var resourceTypes []string
 
-	if dClient != nil {
-		var resourceTypesMap = make(map[string]bool)
-		var resourceTypes []string
-		for _, policy := range policies {
-			for _, rule := range policy.Spec.Rules {
-				for _, kind := range rule.MatchResources.Kinds {
-					resourceTypesMap[kind] = true
-				}
+	for _, policy := range policies {
+		for _, rule := range policy.Spec.Rules {
+			for _, kind := range rule.MatchResources.Kinds {
+				resourceTypesMap[kind] = true
 			}
 		}
+	}
 
-		for kind := range resourceTypesMap {
-			resourceTypes = append(resourceTypes, kind)
-		}
+	for kind := range resourceTypesMap {
+		resourceTypes = append(resourceTypes, kind)
+	}
 
-		resources, err = getResourcesOfTypeFromCluster(resourceTypes, dClient)
+	var resourceMap map[string]map[string]*unstructured.Unstructured
+	if cluster && dClient != nil {
+		resourceMap, err = getResourcesOfTypeFromCluster(resourceTypes, dClient, namespace)
 		if err != nil {
 			return nil, err
 		}
+		if len(resourcePaths) == 0 {
+			for _, rm := range resourceMap {
+				for _, rr := range rm {
+					resources = append(resources, rr)
+				}
+			}
+		}
 	}
 
 	for _, resourcePath := range resourcePaths {
+		lenOfResource := len(resources)
 		resourceBytes, err := getFileBytes(resourcePath)
 		if err != nil {
-			return nil, err
+			if cluster {
+				for _, rm := range resourceMap {
+					for rn, rr := range rm {
+						if rn == resourcePath {
+							resources = append(resources, rr)
+							continue
+						}
+					}
+				}
+			} else {
+				return nil, err
+			}
+			if len(resources) <= lenOfResource {
+				return nil, sanitizedError.NewWithError(fmt.Sprintf("resource %s not found in cluster", resourcePath), errors.New("resource not found"))
+			}
```

Comment:
Why mixing 2 approaches when loading resources? The resources either come from local file or cluster,  this error "resource not found in cluster" seems ambiguous.

Reply:
I have updated the code - separated logic for local file and cluster